### ISSUE TITLE
Decimal number editing in NumbericTextInput

### ISSUE THREAD

**Issue by clemens-msupply** [msupply-foundation/open-msupply#3256]
## What went wrong? 😲

Some numbers are hard/impossible to enter:

- [ ] After entering `1.7` you can't enter `1.701` because `1.70` get immediately reverted to `1.7`
- [ ] Deleting a single digit from the back of (the 2) `1.70002` will immediately result in `1.7` but should be `1.7000`
- [ ] Deleting the `7` from `1.7` is not possible because of the trailing `.`

In general, I think we should try to preserve what the user entered and maybe only format it when the user changes focus otherwise it can become very awkward to edit...

There should be some tests for this behaviour, e.g. I tried to fix these issues but was not sure if I break other things. @CarlosNZ  could you please extend the list above to add test cases you can remember?

- [ ] Write tests for about cases
- [ ] Extend list of possible test cases and include them in the test
- [ ] Fix the issues

**Comment by CarlosNZ** [comment]
Okay, I think those above issues can be boiled down to one thing -- it doesn't handle entering a trailing `0` properly. That should be a fairly simple fix.

**Comment by clemens-msupply** [comment]
Yes got it fixed but required a couple of change so should really add some tests for existing behaviour as well...

**Comment by roxy-dao** [comment]
Can you put this into 1.7 release?

**Comment by CarlosNZ** [comment]
> * Deleting the `7` from `1.7` is not possible because of the trailing `.`

Can't seem to reproduce this one (in Storybook). Is it due to the specific implementation?

> Yes got it fixed but required a couple of change so should really add some tests for existing behaviour as well...

What kind of tests? I'm not seeing any automated tests for UI input components anywhere. We can add more examples to the Storybook components?

**Comment by clemens-msupply** [comment]
I can't reproduce the last point for `1.` This might be in combination of how it is used in JsonForms...

I can reproduce the other point in storiebook though (you might have to increase the number of digits `decimalLimit={5}` though)

I would say adding a test for the component and testing if it shows the correct string in the component. IMHO we shouldn't really do manual testing in storybook especially if there are a number of case like this.

**Comment by CarlosNZ** [comment]
> I would say adding a test for the component and testing if it shows the correct string in the component. IMHO we shouldn't really do manual testing in storybook especially if there are a number of case like this.

I've fixed the main problem here (and a couple of others I noticed), but not sure how to go about creating a test suite. Any ideas? The handful of UI tests we currently have in the repo don't really suggest anything particularly helpful. @clemens-msupply, maybe you could show me what you had in mind when you're in the office next?

**Comment by Chris-Petty** [comment]
Haven't looked at the branch/proposed changes but I'd thoink that just validating/tweaking the input after changing focus/onBlur like Clemen's suggested would be alright?

**Comment by CarlosNZ** [comment]
> Haven't looked at the branch/proposed changes but I'd thoink that just validating/tweaking the input after changing focus/onBlur like Clemen's suggested would be alright?

Yeah, have added additional check on blur. The tricky thing is that many of the implementations require updating the number on each key stroke, so can't just do it all onBlur. Have improved the logic for this though so it should be working for all cases. Will PR this morn.

**Comment by clemens-msupply** [comment]
> > I would say adding a test for the component and testing if it shows the correct string in the component. IMHO we shouldn't really do manual testing in storybook especially if there are a number of case like this.
> 
> I've fixed the main problem here (and a couple of others I noticed), but not sure how to go about creating a test suite. Any ideas? The handful of UI tests we currently have in the repo don't really suggest anything particularly helpful. @clemens-msupply, maybe you could show me what you had in mind when you're in the office next?

Thats an intitial test I came up with:

```tsx
import React, { useState } from 'react';
import { fireEvent, render } from '@testing-library/react';
import { NumericTextInput } from './NumericTextInput';
import { TestingProvider } from '@common/utils';

describe('TestNumericTextInput', () => {
  const TestNumericTextInput = () => {
    const [value, setValue] = useState<number | undefined>(5.0);

    return (
      <TestingProvider>
        <NumericTextInput
          decimalLimit={10}
          value={value}
          onChange={value => setValue(value)}
        />
      </TestingProvider>
    );
  };

  it('Should do something...', async () => {
    const { getByDisplayValue } = render(<TestNumericTextInput />);

    // get by initial value
    const input = getByDisplayValue(5) as HTMLInputElement;

    fireEvent.change(input, { target: { value: '1.7001' } });
    expect(input.value).toBe('1.7001');

    fireEvent.change(input, { target: { value: '1.700' } });
    expect(input.value).toBe('1.700');
  });
});
```

However, there is still some problem in `packages/common/src/intl/number/IntlNumber.ts` where I had to change the `maximumFractionDigits` value to make the component work. Maybe some language (prop 'us' in the test) specific problem?

**Comment by CarlosNZ** [comment]
Closed by #3265
